websocket/src/http_server.cpp

```diff
@@ -0,0 +1,273 @@
+        }
+    }
+
+    static Result Connect(Server* server, uint16_t port)
+    {
+        dmSocket::Socket socket = dmSocket::INVALID_SOCKET_HANDLE;
+        dmSocket::Address bind_address;
+        dmSocket::Result r = dmSocket::RESULT_OK;
+
+        Disconnect(server);
+
+        r = dmSocket::GetHostByName("0.0.0.0", &bind_address);
+        if (r != dmSocket::RESULT_OK)
+        {
+            return RESULT_SOCKET_ERROR;
+        }
+
+        r = dmSocket::New(bind_address.m_family, dmSocket::TYPE_STREAM, dmSocket::PROTOCOL_TCP, &socket);
+        if (r != dmSocket::RESULT_OK)
+        {
+            return RESULT_UNKNOWN;
+        }
+
+        dmSocket::SetReuseAddress(socket, true);
+
+        r = dmSocket::Bind(socket, bind_address, port);
+        if (r != dmSocket::RESULT_OK)
+        {
+            dmSocket::Delete(socket);
+            return RESULT_SOCKET_ERROR;
+        }
+
+        r = dmSocket::Listen(socket, 32);
+        if (r != dmSocket::RESULT_OK)
+        {
+            dmSocket::Delete(socket);
+            return RESULT_SOCKET_ERROR;
+        }
+
+        dmSocket::Address address;
+        uint16_t actual_port;
+        r = dmSocket::GetName(socket, &address, &actual_port);
+        if (r != dmSocket::RESULT_OK)
+        {
+            dmSocket::Delete(socket);
+            return RESULT_SOCKET_ERROR;
+        }
+
+        server->m_Address = address;
+        server->m_Port = actual_port;
+        server->m_ServerSocket = socket;
+
+        return RESULT_OK;
+    }
+
+    Result New(const NewParams* params, uint16_t port, HServer* server)
+    {
+        *server = 0;
+
+        if (!params->m_HttpRequest)
+            return RESULT_ERROR_INVAL;
+
+        Server* ret = new Server();
+        if (Connect(ret, port) != RESULT_OK)
+        {
+            delete ret;
+            return RESULT_SOCKET_ERROR;
+        }
+
+        ret->m_HttpRequest = params->m_HttpRequest;
+        ret->m_Userdata = params->m_Userdata;
+        ret->m_ConnectionTimeout = params->m_ConnectionTimeout * 1000000U;
+        ret->m_Connections.SetCapacity(params->m_MaxConnections);
+
+        *server = ret;
+        return RESULT_OK;
+    }
+
+    void Delete(HServer server)
+    {
+        for (int32_t i; i < server->m_Connections.Size(); ++i)
+        {
+            dmSocket::Shutdown(server->m_Connections[i].m_Socket, dmSocket::SHUTDOWNTYPE_READWRITE);
+            dmSocket::Delete(server->m_Connections[i].m_Socket);
+            server->m_Connections[i].m_Socket = dmSocket::INVALID_SOCKET_HANDLE;
+        }
+        server->m_Connections.SetSize(0);
+        dmSocket::Shutdown(server->m_ServerSocket, dmSocket::SHUTDOWNTYPE_READWRITE);
+        dmSocket::Delete(server->m_ServerSocket);
+        server->m_ServerSocket = dmSocket::INVALID_SOCKET_HANDLE;
+        delete server;
+    }
+
+    /*
+     * Handle an http-connection
+     * Returns 0 - do nothing, 1 - remove, 2 - close.
```

Comment:
Could we add some constants for these?

Reply:
I don't know, this file shouldn't be part of this PR.
Instead we should expose the correct headers from dmSDK.